### PR TITLE
proposal: oneshot login

### PR DESCRIPTION
**Pull request by maxtaco**
- allow a "oneshot" login for keybase, which means that you can specify a username and a paperkey, and the service in question will act on behalf of that user for the duration of its run, acting like a full keybase node
- it's a little bit a hack, in that the device will be the "paper key", but it appears to work in preliminary tests
- we had a request for such a feature to power kbfsgit from inside of short-lived dockers, and it doesn't make sense to provision them as a full keybase devices

### CODE REVIEW

**Comment by maxtaco** [comment]
cc: @zapu, @mmaxim, @mlsteele

**Comment by maxtaco** [comment]
Guys, I threw this together, open to a discussion about it.  If we think it might make it, I'll put together a test or two.  I was able to test it by hand successfully.

**Comment by maxtaco** [comment]
Intended use:

```bash
$ keybase service &
$ KEYBASE_USERNAME=foo KEYBASE_PAPERKEY="ball plane foo bar bam boop" keybase oneshot
```

**Comment by maxtaco** [comment]
Tested to work with git and KBFS headless...

**Comment by bootstraponline** [comment]
I confirm this unblocks using encrypted git on [bitrise's docker containers.](https://github.com/bitrise-docker/android) I'd love to see this pull request merged.

**Comment by maxtaco** [comment]
@bootstraponline this came up during review --- what's the best way to pass secrets into a container?  We're not thrilled about passing paper keys via environment variables.

**Comment by bootstraponline** [comment]
> this came up during review --- what's the best way to pass secrets into a container? We're not thrilled about passing paper keys via environment variables.

Bitrise has a secure process for passing secrets:
- https://devcenter.bitrise.io/bitrise-cli/secrets/

image

I do think that's the right answer. Almost all the CI solutions interface with environment variables and perform some magic to make them secure. Travis CI, for example, supports [encryption](https://docs.travis-ci.com/user/environment-variables/#Defining-encrypted-variables-in-.travis.yml).

File: protocol/avdl/keybase1/login.avdl [hunk 1]
```diff
@@ -81,4 +81,11 @@ protocol login {
     accountDelete is for devel/testing to delete the current user's account.
     */
   void accountDelete(int sessionID);
+
+  /**
+   loginOnetime allows a service to have a "onetime login", without
```
Comment (patrickxb) [inline protocol/avdl/keybase1/login.avdl]: loginOnetime -> loginOneshot

File: go/client/cmd_oneshot.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+flags or the environment.
+
+Passing a paperkey via the environment or via command line flags is
+potentially unsafe. Other processes running on the machine can inspect these
+data, so "oneshot" is strongly advised against on a multi-tenant system where
+users can examine each other's processes. But it might be a good fit for
+Docker deployments.
+
+Also note that by default keybase shouldn't be run as root, but in Docker (or
+other containers), root is the only option, so you can use "keybase oneshot"
```
[the remark this answers is not in the document]
Comment (bootstraponline) [reply go/client/cmd_oneshot.go]: I think it's more true on Docker as a service providers like Bitrise. 😄

**Comment by maxtaco** [comment]
ok @patrickxb for another look

**Pull request merged**